### PR TITLE
Added battle tests for Embargo

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
The first of many milestones required for #3524.
This PR introduces battle tests for the move Embargo, written by none other than yours truly.

> [0] Embargo blocks the effect of an affected Pokémon's held item: PASS
> [0] Embargo blocks an affected Pokémon's trainer from using items: KNOWN_FAILING
> [0] Embargo doesn't block held item effects that affect experience gain: PASS
> [0] Embargo doesn't block held item effects that affect effort values: PASS
> [0] Embargo negates a held item's Speed reduction: PASS
> [0] Embargo doesn't block held item effects that affect friendship: KNOWN_FAILING
> [0] Embargo doesn't block a held item's form-changing effect, but it does any other effects: PASS
> [0] Embargo makes Fling and Natural Gift fail: PASS
> [0] Embargo doesn't stop an item flung at an affected target from activating: PASS
> [0] Embargo is passed via Baton Pass: KNOWN_FAILING
> [0] Embargo doesn't block the effects of berries obtained through Bug Bite or Pluck: PASS
> [0] Embargo disables the effect of the Plate items on the move Judgment: PASS
> [0] Embargo disables the effect of the Drive items on the move Techno Blast: PASS
> [0] Embargo disables the effect of the Memory items on the move Multi Attack: PASS
> [0] Embargo can be reflected by Magic Coat: PASS
> [0] Embargo doesn't prevent Mega Evolution: KNOWN_FAILING
> [0] Embargo doesn't prevent Primal Reversion: PASS
> Please remove KNOWN_FAILING if this test intentionally PASSes
> - Tests PASSED:        13
> - Tests KNOWN_FAILING: 4
> - Tests TOTAL:         17

**Note**: I slapped a `KNOWN_FAILING` to "Embargo doesn't prevent Primal Reversion" because while the test passes, it's not for the right reasons. For more information, see #3554.

## **Discord contact info**
lunos4026

### CODE REVIEW

**Comment by LOuroboros** [comment]
CI was failing because it didn't like that I slapped a "KNOWN_FAILING" to a test that was seemingly working *(for the wrong reasons)*, so I just extended the test a little which makes it fail naturally.

Ready for review.

File: test/battle/move_effect/embargo.c [hunk 1]
```diff
@@ -0,0 +1,394 @@
+#include "global.h"
+#include "test/battle.h"
```
Comment (DizzyEggg) [inline test/battle/move_effect/embargo.c:2]: I would add an Assumption before the tests, that `gBattleMoves[MOVE_EMBARGO].effect == EFFECT_EMBARGO`

File: test/battle/move_effect/embargo.c [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+    PARAMETRIZE { item = ITEM_LUCKY_EGG; }
+    PARAMETRIZE { item = ITEM_NONE; }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { Level(20); Item(item); }
+        OPPONENT(SPECIES_CATERPIE) { Level(10); HP(1); }
+        ASSUME(gItems[ITEM_LUCKY_EGG].holdEffect == HOLD_EFFECT_LUCKY_EGG);
+    } WHEN {
+        TURN { MOVE(player, MOVE_EMBARGO); }
```
Comment (DizzyEggg) [inline test/battle/move_effect/embargo.c:55]: Shouldn't it be Caterpie who's using Embargo on player's Wobbuffet?

File: test/battle/move_effect/embargo.c [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { Item(ITEM_POWER_WEIGHT); }
+        OPPONENT(SPECIES_CATERPIE) { HP(1); }
+        ASSUME(gItems[ITEM_POWER_WEIGHT].holdEffect == HOLD_EFFECT_POWER_ITEM);
+        ASSUME(gItems[ITEM_POWER_WEIGHT].holdEffectParam == 8);
+        ASSUME(gItems[ITEM_POWER_WEIGHT].secondaryId == STAT_HP);
+        ASSUME(gSpeciesInfo[SPECIES_CATERPIE].evYield_HP == 1);
+    } WHEN {
+        TURN { MOVE(player, MOVE_EMBARGO); }
```
Comment (DizzyEggg) [inline test/battle/move_effect/embargo.c:82]: The same here, shouldn't it be opponent using Embargo on player's mon?

File: test/battle/move_effect/embargo.c [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    PARAMETRIZE { heldItem = ITEM_NONE; }
+    PARAMETRIZE { heldItem = ITEM_MEADOW_PLATE; }
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_ARCEUS_GRASS) { Item(heldItem); };
+        ASSUME(gItems[ITEM_MEADOW_PLATE].holdEffect == HOLD_EFFECT_PLATE);
+        ASSUME(gItems[ITEM_MEADOW_PLATE].holdEffectParam == 20);
+        ASSUME(gItems[ITEM_MEADOW_PLATE].secondaryId == TYPE_GRASS);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_RAZOR_LEAF); }
```
Comment (DizzyEggg) [inline test/battle/move_effect/embargo.c:152]: Embargo is not used at all here

File: test/battle/move_effect/embargo.c [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        // Turn 2
+        if (moveId == MOVE_FLING)
+            MESSAGE("Wobbuffet used Fling!");
+        else
+            MESSAGE("Wobbuffet used Natural Gift!");
+        MESSAGE("But it failed!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Embargo doesn't stop an item flung at an affected target from activating")
```
Comment (DizzyEggg) [inline test/battle/move_effect/embargo.c:185]: I don't get this test. What is exactly being checked here? Player uses Embargo on opponent, and then uses Fling.

**Review by DizzyEggg — DISMISSED** [review]

File: test/battle/move_effect/embargo.c (hunk 5 above)
Comment (LOuroboros) [reply test/battle/move_effect/embargo.c:185]: It tests that the effect of the item that is flung at a target that is suffering the effects of Embargo activates normally 😄

File: test/battle/move_effect/embargo.c (hunk 2 above)
Comment (LOuroboros) [reply test/battle/move_effect/embargo.c:55]: Yup. I'll correct it!

File: test/battle/move_effect/embargo.c (hunk 3 above)
Comment (LOuroboros) [reply test/battle/move_effect/embargo.c:82]: Yup. I'll correct it!

**Comment by LOuroboros** [comment]
Ready for re-review :eyes:

File: test/battle/move_effect/embargo.c (hunk 5 above)
Comment (LOuroboros) [reply test/battle/move_effect/embargo.c:186]: In other words, it verifies that Embargo is only blocking the affected Pokémon's own item effects, and that they're not blocking item effects that would affect them.

I don't know if it's a silly test to make or not, I just wrote tests based on what's written at places such as Smogon and Bulbapedia 😅 ...

Comment (Bassoonian) [reply test/battle/move_effect/embargo.c:186]: Doesn't that mean we need a config for Gen IV behaviour then?

Comment (LOuroboros) [reply test/battle/move_effect/embargo.c:186]: This is what I was told on Discord when I was still writing the Embargo and Mega Evolution test:

> SBird (Karathan) 11/11/2023 8:56 AM
> the tests can't currently override preproc constants
> (I think)

:eyes:

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**